### ISSUE TITLE
ModelRenderRegistry.addItemToRegistry with custom ModID

### ISSUE THREAD

**Issue by Kovak** [ObliviousSpartan/SpartanWeaponry#1]
It looks like at the moment it expects all item model files to be under the spartanweaponry domain. It would be nice if we could specify a modid in addItemToRegistry so that our custom models could be under their own mod id.

**Comment by ObliviousSpartan** [comment]
Sorry for the late response. I need to set up notifications for issues received.
How is it that you are registering your item models? Because it looks like you are using the internal Spartan Weaponry method of doing this. That hasn't been exposed as part of the API.

I can provide a method to put in your model registration class if you like that you can call in the same class to put the correct model files in your mod domain. You will need to however provide the model names yourself to use it.
```
        public static void registerItemRender(Item item, String modelName)
	{
		ModelLoader.setCustomModelResourceLocation(item, 0, new ModelResourceLocation(modelName, "inventory"));
	}

Let me know if this helps.

**Comment by Kovak** [comment]
Oh yes I have been using the internal model registration. Something like:

```
Item katana = SpartanWeaponryAPI.createKatana(
                    mat.material,
                    SpartanFire.MODID,
                    DamageHelper.getDamage(DamageHelper.WeaponType.KATANA,
                            mat.material.getMaterial().getAttackDamage()),
                    CreativeTabsSW.TAB_SW
                    );
ModelRenderRegistry.addItemToRegistry(katana, "katana_" + mat.material.getUnlocName());
item_set.add(katana);
```

How do you envision us registering our models?

**Comment by Kovak** [comment]
I think my example I provided didn't fully summarize the problem since it uses a custom texture.

If I'm going to use the default ToolMaterialEx color property based rendering, I've been calling:

`
ModelRenderRegistry.addItemToRegistry(throwing_axe,"throwing_axe_" + mat.material.getUnlocName(),
                    mat.material);
`

How are we to use the ColourHandler system if these parts aren't exposed?

**Comment by ObliviousSpartan** [comment]
That has been directly exposed as part of the API now. See below:

    SpartanWeaponryAPI.registerColourHandler(Item... items)

**Comment by Kovak** [comment]
oh i didnt notice that, I see so we we should have our own render registry but then use the colour handler

**Comment by ObliviousSpartan** [comment]
I'll have to see about doing that. I'll also will have to expose it to the API itself, so you wouldn't have to decompile the mod itself. I'll see what I can do.

**Comment by ObliviousSpartan** [comment]
I have updated the API on GitHub, adding a way to register the Item model, either directly, only requiring the Item class, or another version that requires the Item class, your Mod ID, and the model name. You will need version "beta-1.2.1" of Spartan Weaponry (currently pending approval) to use it however. Hopefully that helps you out.

**Comment by Kovak** [comment]
Thanks so much

**Comment by Kovak** [comment]
Sorry to re-open this I just got around to trying to update another one of the compat mods.

While you exposed registerItemRender with a modID api, this is actually not the part that was most helpful imo. My idea was allowing the compat mods to piggy-back off your registry in the first place making it so they do not need to include any registries other than the items themselves by changing:

```java
    @SubscribeEvent
    public static void registerItemRenders(ModelRegistryEvent ev) {
        registerMultiItemRender(ItemRegistrySW.material);
        Iterator var1 = models.iterator();

        while(var1.hasNext()) {
            ModelRenderRegistry.ModelRegistryEntry entry = (ModelRenderRegistry.ModelRegistryEntry)var1.next();
            registerItemRender(entry.getItem(), "spartanweaponry", entry.getModelLocation());
        }

        models.clear();
        LogHelper.info("Registered all item renders!");
    }
}
```

to

```java
    @SubscribeEvent
    public static void registerItemRenders(ModelRegistryEvent ev) {
        registerMultiItemRender(ItemRegistrySW.material);
        Iterator var1 = models.iterator();

        while(var1.hasNext()) {
            ModelRenderRegistry.ModelRegistryEntry entry = (ModelRenderRegistry.ModelRegistryEntry)var1.next();
            registerItemRender(entry.getItem(), entry.getModId(), entry.getModelLocation());
        }

        models.clear();
        LogHelper.info("Registered all item renders!");
    }
}
```

and then there would be method variants like:

`
public static void addItemToRegistry(Item item, String modelLoc, String modId) 
public static void addItemToRegistry(Item item, String modelLoc, ToolMaterialEx material, String modId) 
`

This makes them lean, mean, and means they don't have to duplicate the whole ModelRegistryEntry structure or worry about getting their rendering initialization exactly right (especially when its reading/copying half-obfuscated code). 

At the moment all my compat mods will have to have a complete copying of your ModelRenderRegistry to just swap out the hardcoded modid provided for the location of the models. It would be a lot easier to maintain these mods if all they had to do was register their items and pass them to your SpartanWeaponry's own ModelRenderRegistry so that code doesn't become sharded across all the implementors. All of this already works if you use the current endpoints and put the model json under spartanweaponry instead of the modid.

**Comment by ObliviousSpartan** [comment]
Sorry about the late response. I've just moved house recently and got internet set up again. Unfortunately it is rather crappy internet than what I'm used to.

Anyway, just for clarity's sake, is the way you want this to be done is so that Spartan Weaponry adds the model entry to the mod's internal registry, and then let Spartan Weaponry do all the registration internally, rather than directly calling the registration function? If I've read the above correctly, I believe that's what you want to do right?

**Comment by Kovak** [comment]
Yes my goal is that I can call addItemToRegistry, like I was doing here with the old version:

https://github.com/ChaosBuffalo/Spartan-BaseMetals/blob/master/src/main/java/com/chaosbuffalo/spartanbm/init/ItemRegistrySBM.java#L58

But be able to provide my own modid so that the model files can be kept in the right asset directory. This way the compatibility mods only need to have their own item registries and let your ModelRegistry do all the registration.

**Comment by ObliviousSpartan** [comment]
Hi. Sorry for the very late response. I had to deal with a few stressful things in real life that took my time away from my mod development.
Anyway, I've got a progress update for you. What I've done in my internal build of the mod is change how my internal class `ModelRegistryEntry()` holds it's model location. It has been changed from a `String` to a `ResourceLocation` in my current development build. I've done this because the `ResourceLocation` class allows specifying a domain for resource files, which, in most cases is the mod's ID. 
What I intend on doing with this is exposing this method to the API directly, using the following method name and it's overloads
```
AddItemModelToRegistry(Item item)
AddItemModelToRegistry(Item item, String modId, String modelLocation)
AddItemModelToRegistry(Item item, ResourceLocation modelLocation)
```
and mark the previous ones that I made for this issue as deprecated.
This will mean I'll be uploading a new version of the API when that is done. I'll let you know when it's up.
